primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,130 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameter set is encoded as an opaque vector which structure depends on the execution
+//! environment itself (except for environment type/version which is always represented
+//! by the first element of the vector). Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use crate::v2::{BlakeTwo256, HashT as _};
+use parity_scale_codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use parity_util_mem::MallocSizeOf;
+use polkadot_core_primitives::Hash;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec, vec::Vec};
+
+/// # Execution environment parameter tags
+/// Values from 1 to 16 are reserved for system-wide parameters not related to any concrete
+/// execution environment.
+/// Environment type
+pub const EEPAR_01_ENVIRONMENT: u32 = 1;
+/// Logical stack limit, in stack items
+pub const EEPAR_17_STACK_LOGICAL_MAX: u32 = 17;
+/// Native stack limit, in bytes
+pub const EEPAR_18_STACK_NATIVE_MAX: u32 = 18;
+
+/// # Execution enviroment types
+/// Generic Wasmtime environment
+pub const EXEC_ENV_TYPE_WASMTIME_GENERIC: u32 = 0;
+
+/// Unit type wrapper around [`type@Hash`] that represents an execution parameter set hash.
+///
+/// This type is produced by [`ExecutorParams::hash`].
+#[derive(Clone, Copy, Encode, Decode, Hash, Eq, PartialEq, PartialOrd, Ord, TypeInfo)]
+pub struct ExecutorParamsHash(Hash);
+
+impl sp_std::fmt::Display for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		self.0.fmt(f)
+	}
+}
+
+impl sp_std::fmt::Debug for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "{:?}", self.0)
+	}
+}
+
+impl sp_std::fmt::LowerHex for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		sp_std::fmt::LowerHex::fmt(&self.0, f)
+	}
+}
+
+/// # Deterministically serialized execution environment semantics
+/// Represents an arbitrary semantics of an arbitrary execution environment, so should be kept as
+/// abstract as possible. Mapping from `u32` constant tags to SCALE-encoded values was chosen
+/// over an idiomatic `enum` representation as the latter would require caller to rely on
+/// deserialization errors when decoding a structure of newer version not yet known to the node
+/// software, which is undesirable.
```

Comment:
Why is that undesirable? Only retrieving like half the environment seems to be a bad idea from a consensus perspective. In practice we should only role out new values, once they are sufficiently understood by nodes on the network and then any nodes still on the old version, should simply fail. To me that would be safer than trying to interpret half the environment and ignoring the rest.

Reply:
Again, @pepyakin is the final boss to fight here :) His arguments on this topic seemed convincing to me, but it would be better if he could tell about this architectural decision himself, not to misquote him.

Reply:
Yes please. Even if well justified, those decisions are in no way obvious, so I would very much like to have them properly documented in likely the implementer's guide.